Adding new EndpointSlice Mirroring Controller

robscott opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This adds an EndpointSlice mirroring controller to mirror custom Endpoints to EndpointSlices as [described by the KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/20190603-EndpointSlice-API.md#endpointslicemirroring-controller). 

**Does this PR introduce a user-facing change?**:
```release-note
Custom Endpoints are now mirrored to EndpointSlices by a new EndpointSliceMirroring controller.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

- [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/20190603-endpointslices/README.md)
- [Enhancement Issue](https://github.com/kubernetes/enhancements/issues/752) 

**Test Coverage**
Progress on required test coverage from the KEP.

**Unit Tests**
* Stale EndpointSlices managed by EndpointSliceMirroring controller are cleaned up by the controller when it starts up.
  _Covered by controller tests, primarily an issue for garbage collector since these are owned by Endpoints resources._

* Only Endpoints with the skip label set to "true" will be mirrored by the EndpointSliceMirroring controller.
  This has changed a bit, but is _covered by controller tests._

* Endpoints transitioning between values of the skip label should result in corresponding EndpointSlices being created or deleted.
  _Covered by controller tests, could potentially use additional coverage in e2e or integration tests_

* Mirrored EndpointSlices will:
  * Include the appropriate labels, endpoints, and ports of the corresponding Endpoints resource.
    _Covered by unit tests._
    
  * Refer to the original Endpoints resource with an OwnerReference and a kubernetes.io/service-name label.
    _Covered by unit tests._ 

  * Include a endpointslice.kubernetes.io/managed-by label set to endpointslicemirroring-controller.k8s.io.
    _Covered by unit tests._ 

* Endpoints with dual stack Endpoints should be represented with multiple EndpointSlices, at least 1 per IP family.
  _Covered by unit tests, could potentially add some scale with unit or integration tests._

* Endpoints with multiple subsets should be represented with multiple EndpointSlices, at least 1 per subset.
  _Covered by unit tests, could potentially add some scale with unit or integration tests._

* Endpoints with multiple subsets and IP families should be represented with multiple EndpointSlices, at least 1 per subset and IP family.
  _Covered by unit tests, could potentially add some scale with unit or integration tests._

* Endpoints with more than 1000 endpoints should result in exactly 1000 endpoints being mirrored to corresponding EndpointSlices.
  _Ended up with a variation that will limit replication to 1000 endpoints per subset, covered by controller tests._

* EndpointSlices with a different endpointslice.kubernetes.io/managed-by will not be modified by the EndpointSlice mirroring controller.
  _Covered in controller tests._

* When a selector is added to a Service with a preexisting Endpoints resource:
  * ~~The Endpoints controller will modify the existing Endpoints resource to add the endpointslice.kubernetes.io/skip-mirror label.~~
    ~~Covered by unit and integration tests.~~

  * The EndpointSliceMirroring controller will delete any EndpointSlices it has mirrored for that Endpoints resource.
    _EndpointSliceMirroring controller doesn't watch Services, but other tests
    verify that skip label is respected._


**E2E Tests**
* Custom Endpoints are mirrored through create, update, and delete process.
  _Covered by e2e tests._

* ~~Endpoints created by the Endpoints controller include the skip label.~~
  ~~Covered by e2e and integration tests.~~

* EndpointSlices are created with the appropriate labels, endpoints, ports, and Service references (label and owner ref).
  _Covered fully by unit tests, partially by e2e tests._


/sig network
/hold

fedebongio commented:
/cc @bowei

aojea commented:
/cc

robscott commented:
This is not quite ready for review, but it's close. I'm transitioning this from draft status to get tests running.

bowei reviewed (commented):
just started reviewing./..

robscott commented on pkg/controller/endpointslicemirroring/config/types.go:
```diff
@@ -0,0 +1,43 @@
+	// mirroringConcurrentServiceEndpointSyncs is the number of service endpoint
+	// syncing operations that will be done concurrently. Larger number = faster
+	// endpoint slice updating, but more CPU (and network) load.
+	MirroringConcurrentServiceEndpointSyncs int32
```
Sadly these names need to be unique since they are translated into kube-controller-manager flags. I could reuse the values from the EndpointSlice controller flags but it seems like it would be good to allow this controller to be configured separately.

cheftako commented on cmd/kube-controller-manager/app/discovery.go:
```diff
@@ -52,3 +53,24 @@ func startEndpointSliceController(ctx ControllerContext) (http.Handler, bool, er
+
+func startEndpointSliceMirroringController(ctx ControllerContext) (http.Handler, bool, error) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.EndpointSlice) {
+		klog.V(4).Infof("Not starting endpointslicemirroring-controller since EndpointSlice feature gate is disabled")
```
+1

robscott replied:
Updated, thanks!

cheftako commented on hack/.golint_failures:
```diff
@@ -59,6 +59,7 @@ pkg/controller/disruption
 pkg/controller/endpoint
 pkg/controller/endpoint/config/v1alpha1
 pkg/controller/endpointslice/config/v1alpha1
+pkg/controller/endpointslicemirroring/config/v1alpha1
```
Can we just fix the lint errors?

robscott replied:
Unfortunately I think this is a problem with the required conversion functions like `Convert_v1alpha1_EndpointSliceMirroringControllerConfiguration_To_config_EndpointSliceMirroringControllerConfiguration`. As far as I can tell, every controller has an entry for this config in `.golint_failures`.

cheftako commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,471 @@
+		go wait.Until(c.worker, c.workerLoopPeriod, stopCh)
+	}
+
+	go func() {
```
Whats the point in spinning of a separate goroutine which just does a deferred handle crash, especially as this run method already has a deferred handle crash.

robscott replied:
🤦 that was apparently a relic from the Endpoints controller - there was actually a function call in the goroutine. Somehow I managed to remove the function call but left the pointless goroutine. Thanks for catching this one!

robscott commented on pkg/controller/endpointslicemirroring/endpointset.go:
```diff
@@ -0,0 +1,96 @@
+}
+
+// Returns an endpoint matching the hash if contained in the set.
+func (s endpointSet) Get(item *discovery.Endpoint) *discovery.Endpoint {
```
This is actually a direct copy of the set defined by the EndpointSlice controller. I think that may have been based on another common k8s Set datastructure, but can't quite remember. I'm open to changing the signature of this function at some point, but would prefer to do it once we're able to combine/share some of the code for these controllers. For the sake of simpler backporting I'm trying to avoid any changes like that here.

bowei replied:
ok

cheftako commented on cmd/kube-controller-manager/app/discovery.go:
```diff
@@ -52,3 +53,25 @@ func startEndpointSliceController(ctx ControllerContext) (http.Handler, bool, er
 	return nil, true, nil
 }
+
+func startEndpointSliceMirroringController(ctx ControllerContext) (http.Handler, bool, error) {
```
For my paranoia's sake can we add this to the TestController_DiscoveryError test in https://github.com/kubernetes/kubernetes/blob/release-1.18/cmd/kube-controller-manager/app/core_test.go#L100?

robscott replied:
Good call, I've added this and the EndpointSlice controller to that test.

robscott commented on pkg/controller/endpointslicemirroring/endpointset.go:
```diff
@@ -0,0 +1,96 @@
+	return endpointHash(endpointutil.DeepHashObjectToString(hashObj))
+}
+
+// endpointSet provides simple methods for comparing sets of Endpoints.
```
Commented below about wanting to maintain consistency with the EndpointSlice controller for now so we can potentially share this code in the future.

robscott commented:
Thanks to @bowei and @cheftako for the initial reviews! I've been iterating on this PR for quite awhile now and I think it's ready for a full round of review. I believe I've covered everything initially proposed by the EndpointSlice KEP for EndpointSlice mirroring. I've also updated the PR description to cover how/where the proposed test plan has been covered.

/assign @thockin @freehan
/hold cancel
/priority important-soon

bowei commented on pkg/controller/endpointslicemirroring/endpointset.go:
```diff
@@ -0,0 +1,96 @@
+// should attributes such as topology, conditions, or targetRef change.
+type endpointHash string
+type endpointHashObj struct {
+	Addresses []string
```
is Addresses sorted?

robscott replied:
Good question - it is sorted as part of the `hashEndpoint()` function that's called for all major operations: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpointslice/endpointset.go#L36.

robscott commented:
/retest

bowei commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,405 @@
+// EndpointSlices exist. It creates, updates, or deletes EndpointSlices to
+// ensure the desired set of addresses are represented by EndpointSlices.
+func (r *reconciler) reconcile(endpoints *corev1.Endpoints, existingSlices []*discovery.EndpointSlice) error {
+	slicesToCreate := []*discovery.EndpointSlice{}
```
You don't reference these until much further down in the code. Move to before use.
Also, you can do to avoid having multiple similar "things" floating around

```
slices := struct {
  toCreate, toUpdate, toDelete []*discovery.EndpointSlice
}
```

robscott replied:
Good call - I've moved to a new struct type to organize slices by action and updated all uses of that pattern to use the new type.

bowei commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,405 @@
+	slicesToDelete := []*discovery.EndpointSlice{}
+
+	// Build data structures for existing state.
+	existingSlicesByKey := map[addrTypePortMapKey][]*discovery.EndpointSlice{}
```
Is this only referenced much further below? If possible, move closer to usage

robscott replied:
Good call, moved further down.

bowei commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,405 @@
+	}
+
+	// Build data structures for desired state.
+	desiredPortsByKey := map[addrTypePortMapKey][]discovery.EndpointPort{}
```
It feels like the below loop can be made easier to understand (and test) with a small utility object:

```
type desired struct {
  portsByKey map[addrTypePortMapKey][]discovery.EndpointPort{}
  endpointsByKey map[addrTypePortMapKey]endpointSet{}
  numDesiredEndpoints int
}
 
func (d*desired) updatePorts(endpointPorts) { ... }
func (d*desired) addAddr(a) { ... }
func (d*desired) addNotReadyAddr(a) { ... }
```

then the loop becomes:

```
d := newDesired()
for _, subset := range endpoints.Subsets {
  d.updatePorts(epPortsToEpsPorts...)
  for _, a := range subset.Addresses {
    d.addAddress(a)
  }
  for _, a := range subset.NotReadyAddresses {
    d.addNotReadyAddr(a)
  }
}
```

robscott replied:
Good idea. It didn't turn out quite as elegant thanks to some additional complexity, but I think the end result is still an improvement.

bowei commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,405 @@
+			if addrType == nil {
+				klog.Warningf("Address in %s/%s Endpoints is not a valid IP, it will not be mirrored to an EndpointSlice: %s", endpoints.Namespace, endpoints.Name, address.IP)
+			} else {
+				epHash := v4Hash
```
create a helper object to make this easier to understand

```
type keyer struct { // probably a better name for this
  v4, v6 addrTypePortMapKey
}

func newKeyer(endpointPorts) *keyer { ... }

func (k* keyer) forAddress(addr) ... {
}

desiredEndpointsByKey[k.forAddress(address)].Insert(endpoint)
```

bowei commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,405 @@
+	}
+
+	epMetrics := mirroringmetrics.NewEndpointPortCache()
+	totalAdded := 0
```
total := struct { added, removed int }

robscott replied:
Good idea, updated all uses of that pattern to use this kind of struct.

bowei commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,405 @@
+		})
+
+		if len(pmSlicesToCreate) > 0 {
+			slicesToCreate = append(slicesToCreate, pmSlicesToCreate...)
```
if the `if` required? append of empty slice is nop?

robscott replied:
Updated, thanks!

bowei reviewed (commented):
Some recommendations for simplifying the code with a few well placed objects

bowei commented:
/assign

bowei commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,405 @@
+	sliceNamesToDelete := sets.String{}
+	numRemoved := 0
+
+	// 1. Iterate through existing slices to delete endpoints no longer desired
```
It feels like each discrete step should be broken up into methods on a helper object (like the desired struct example above),
then it will be easier to understand

bowei replied:
```
type sliceInfo struct {
  eps *EndpointSlice
  originalLen int
  updated bool
}

func (si*sliceInfo) update(endpoints) {
  if len(endpoints) == 0 { .. mark for delete }
  if ...
}

type slices map[/*slice.Name*/string]*sliceinfo

all := slices{}
for _, eps := range existingSlices {
  all[s.Name] = &sliceInfo{eps, len(eps.Endpoints), false}
}

for _, si := range all {
  newEps, changed := computeEPSUpdate(all, ...)
  if updated || len(...) { si.update(newEps) }
}

// etc...
```

robscott replied:
I ended up trying a few ways to clean this up and it all ended up being messier than necessary. Instead I've gone back to my original plan which was to have a 1:1 mapping between EndpointSubset and EndpointSlice and remove the vast majority of this logic. Since we were already limiting the endpoints mirrored per subset to 1000, those can call fit in a single EndpointSlice. I can't imagine custom endpoints being a common use case beyond that, and if they are, it's likely time to switch to using the EndpointSlice API. With that said, I know this is a bit of a controversial change, so I've added this as an additional commit for now so it can easily be rolled back.

robscott replied:
I've gone all in with this approach, I think it's cleaner and simpler. Initially I'd hoped that we could share more logic between this and the EndpointSlice controller, but it doesn't really make sense to add complexity here for the sake of uniformity with the other controller. I've integrated this approach into the original set of commits.

bowei commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,405 @@
+				// otherwise, copy and mark for update
+				epSlice := existingSlice.DeepCopy()
+				epSlice.Endpoints = newEndpoints
+				slicesByName[existingSlice.Name] = epSlice
```
updating the slicesByName in multiple places in this code makes me nervous about messing up the logic.

robscott replied:
Potentially resolved with last commit - see comment above for more context.

bowei commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,405 @@
+			}
+			if len(newEndpoints) == 0 {
+				// if no endpoints desired in this slice, mark for deletion
+				sliceNamesToDelete.Insert(existingSlice.Name)
```
can we take care of this at the very end somehow (e.g. loop through all slices and those that end up with no endpoints will be marked for deletion)

robscott replied:
Potentially resolved with last commit - see comment above for more context.

aojea commented on test/e2e/network/endpointslicemirroring.go:
```diff
@@ -0,0 +1,222 @@
+					LabelSelector: discoveryv1beta1.LabelServiceName + "=" + svc.Name,
+				})
+				if err != nil {
+					return false, err
```
does this retry or fail?

robscott replied:
Good call, I hadn't thought an API error here would be very likely but still very much something I should retry for, updated.

aojea commented on test/e2e/network/endpointslicemirroring.go:
```diff
@@ -0,0 +1,222 @@
+			_, err := cs.CoreV1().Endpoints(f.Namespace.Name).Create(context.TODO(), endpoints, metav1.CreateOptions{})
+			framework.ExpectNoError(err, "Unexpected error creating Endpoints")
+
+			if err := wait.PollImmediate(2*time.Second, 12*time.Second, func() (bool, error) {
```
don't know if 12 secs will be too short for a busy CI

robscott commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,465 @@
+	// represent the sequence of delays between successive queuings of an
+	// Endpoints resource.
+	//
+	// 1s, 2s, 4s, 8s, 16s, 32s, 64s, 128s, 256s, 512s, 1000s (max)
```
Good call, this was confusing. The 1000s was representing the upper limit of delays (configured by `maxSyncBackOff`). That had been decreased to 100s in the EndpointSlice controller without a corresponding update to the comment. Fixing that here.

robscott commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,465 @@
+
+	// maxEndpointsPerSubset is the maximum number of endpoints that will be
+	// mirrored from an EndpointSubset.
+	maxEndpointsPerSubset = 1000
```
This is actually distinct from the `maxEndpointsPerSlice` that can be configured with a corresponding `kube-controller-manager` flag and has the same default value of 100. This was an arbitrary number introduced to ensure we weren't mirroring huge numbers of endpoints. At a certain scale, it likely makes more sense for users of custom endpoints to be writing EndpointSlices instead of relying on this mirroring mechanism. This is a relatively arbitrary value that lines up with the max size for an EndpointSlice so we could theoretically achieve a 1:1 mapping between EndpointSubsets and EndpointSlices, but that's not actually what ended up happening here.

robscott replied:
I've cleaned this up so that `maxEndpointsPerSubset` is configurable as a controller flag and replaces the `maxEndpointsPerSlice` altogether.

robscott commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,465 @@
+	klog.V(2).Infof("Waiting for application of %s label", discovery.LabelSkipMirror)
+	c.waitForSkipLabelApplication()
+	klog.V(2).Infof("%s label successfully applied", discovery.LabelSkipMirror)
+
```
Updated, thanks!

robscott commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,465 @@
+// has been applied to Endpoints managed by the Endpoints controller. This
+// function waits until that's true.
+func (c *Controller) waitForSkipLabelApplication() error {
+	return wait.PollImmediate(2*time.Second, 5*time.Minute, func() (bool, error) {
```
Updated, thanks!

robscott commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,465 @@
+		}
+
+		if numEndpointsWithProblems > 0 {
+			klog.V(2).Infof("Waiting to start EndpointSliceMirroring controller until Endpoints controller applies %s label (waiting for %d/%d Endpoints)", discovery.LabelSkipMirror, numEndpointsWithProblems, len(endpoints))
```
Good call, updated.

robscott commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,465 @@
+// function waits until that's true.
+func (c *Controller) waitForSkipLabelApplication() error {
+	return wait.PollImmediate(2*time.Second, 5*time.Minute, func() (bool, error) {
+		endpoints, err := c.endpointsLister.List(labels.Everything())
```
Good call, this refactoring + testing caught another bug - I wasn't properly handling identical names in different namespaces.

robscott commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,465 @@
+	}
+
+	klog.V(2).Infof("Waiting for application of %s label", discovery.LabelSkipMirror)
+	c.waitForSkipLabelApplication()
```
Done, thanks!

robscott commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,465 @@
+	}
+
+	if c.queue.NumRequeues(key) < maxRetries {
+		klog.Warningf("Error mirroring EndpointSlices for %q Endpoints, retrying. Error: %v", key, err)
```
I've moved to this approach because it reads more naturally to me to have the name followed by the type ("example for foo/bar Endpoints" vs "example for Endpoints foo/bar". I think it's just that I think the name is more important than the type here, or maybe it's the weird pluralization of Endpoints? Not quite sure why I like having the name first, but I'm open to changing it.

robscott commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,465 @@
+func (c *Controller) syncEndpoints(key string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing EndpointSlices for %q Endpoints. (%v)", key, time.Since(startTime))
```
That would be a great metric to add, don't think it would be too difficult. Should likely add this to EndpointSlice controller as well.

robscott replied:
Just added a metric to track the duration of syncEndpoints() calls. Unfortunately all the test cases take 0ms so it's not easy to reliably test, but I did a one off test case where I added a `time.Sleep()` call in `syncEndpoints` and the metric worked as expected.

robscott commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,465 @@
+	defer func() {
+		klog.V(4).Infof("Finished syncing EndpointSlices for %q Endpoints. (%v)", key, time.Since(startTime))
+	}()
+
```
I've added this but I'm a little concerned over the potential noise since it's mostly duplicating the log message above. In practice the log message above is already relatively spammy in kube-controller-manager. On the other hand, if/when things really break it's helpful to know that we at least reached the syncEndpoints() function.

bowei replied:
What is the standard log level of the controller? V(4) expectation is log spam..

robscott replied:
That's fair. OSS clusters default to v(4) but that's likely rarely used for prod clusters.

robscott commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,465 @@
+func (c *Controller) queueEndpoints(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %+v: %v", obj, err))
```
Updated, thanks!

robscott commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,465 @@
+func (c *Controller) onEndpointsAdd(obj interface{}) {
+	endpoints := obj.(*v1.Endpoints)
+	if endpoints == nil {
+		utilruntime.HandleError(fmt.Errorf("Invalid Endpoints provided to onEndpointsAdd()"))
```
Updated, thanks!

robscott commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,465 @@
+}
+
+// onEndpointsAdd queues a sync for the relevant Endpoints resource if
+// skipMirror() is not true.
```
Updated, thanks!

robscott commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,465 @@
+	endpoints := obj.(*v1.Endpoints)
+	prevEndpoints := prevObj.(*v1.Endpoints)
+	if endpoints == nil || prevEndpoints == nil {
+		utilruntime.HandleError(fmt.Errorf("Invalid Endpoints provided to onEndpointsUpdate()"))
```
Updated, thanks!

robscott commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,465 @@
+		return
+	}
+	if (skipMirror(endpoints.Labels) && skipMirror(prevEndpoints.Labels)) || forLeaderElection(endpoints.Annotations) {
+		return
```
I think logging at v(4) could might drown out other more important logs at the same level. Since I'm assuming we'll be ignoring significantly more Endpoints than we're mirroring, often 100%. I've added logging at v(5), but open to feedback on that.

robscott commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,465 @@
+		return
+	}
+	if skipMirror(endpoints.Labels) || forLeaderElection(endpoints.Annotations) {
+		return
```
I think logging at v(4) could might drown out other more important logs at the same level. Since I'm assuming we'll be ignoring significantly more Endpoints than we're mirroring, often 100%. I've added logging at v(5), but open to feedback on that.

bowei replied:
Is the controller really logging at V(4)? What is the meaning of levels 2, 3 then?

My expectations:

V(1) - Always logged 
V(2) - Major lifecycle, should not impact performance
V(3) - Minor steps, spammy but not peformance impacting
V(4) - Very spammy, may be a bad idea under load
V(5) - ??

robscott commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,465 @@
+func (c *Controller) onEndpointSliceAdd(obj interface{}) {
+	endpointSlice := obj.(*discovery.EndpointSlice)
+	if endpointSlice == nil {
+		utilruntime.HandleError(fmt.Errorf("Invalid EndpointSlice provided to onEndpointSliceAdd()"))
```
Updated, thanks!

robscott commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,465 @@
+		utilruntime.HandleError(fmt.Errorf("Invalid EndpointSlice provided to onEndpointSliceUpdate()"))
+		return
+	}
+	if managedByChanged(prevEndpointSlice, endpointSlice) || (managedByController(endpointSlice) && c.endpointSliceTracker.Stale(endpointSlice)) {
```
If the controller restarts all Endpoints would naturally get resynced so I think that would handle this. The reason this exists is for the edge case that someone may change an EndpointSlice managed by this controller to be managed by something else and this controller would need to catch that event and resync/add a new EndpointSlice.

bowei replied:
Does the code cover changed managed by + restart in the middle of that operation?

robscott commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,465 @@
+	// queue after the max of endpointSliceChangeMinSyncDelay and
+	// endpointUpdatesBatchPeriod.
+	delay := endpointSliceChangeMinSyncDelay
+	if c.endpointUpdatesBatchPeriod > delay {
```
Good catch, this was a relic from the EndpointSlice controller that had slightly more complicated logic that's no longer needed here. Should be fixed now.

robscott commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,465 @@
+		return err
+	}
+
+	esLabelSelector := labels.Set(map[string]string{
```
Good idea - I've moved this to a new function and added some unit tests.

robscott commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,405 @@
+	slicesToDelete := []*discovery.EndpointSlice{}
+
+	// Build data structures for existing state.
+	existingSlicesByKey := map[addrTypePortMapKey][]*discovery.EndpointSlice{}
```
Good idea, moved to a function.

robscott commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,405 @@
+			endpoint := addressToEndpoint(address, true)
+			addrType := getAddressType(address.IP)
+			if addrType == nil {
+				klog.Warningf("Address in %s/%s Endpoints is not a valid IP, it will not be mirrored to an EndpointSlice: %s", endpoints.Namespace, endpoints.Name, address.IP)
```
Good call - updated to add a warning event below the loops.

robscott commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,405 @@
+
+	for _, subset := range endpoints.Subsets {
+		endpointPorts := epPortsToEpsPorts(subset.Ports)
+		v4Hash := newAddrTypePortMapKey(endpointPorts, discovery.AddressTypeIPv4)
```
Good catch, thanks!

robscott commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,405 @@
+			endpoint := addressToEndpoint(address, false)
+			addrType := getAddressType(address.IP)
+			if addrType == nil {
+				klog.Warningf("Address in %s/%s Endpoints is not a valid IP, it will not be mirrored to an EndpointSlice: %s", endpoints.Namespace, endpoints.Name, address.IP)
```
Good call - updated to add a warning event below the loops.

bowei commented on pkg/controller/endpointslicemirroring/endpointslice_tracker.go:
```diff
@@ -0,0 +1,123 @@
+	}
+}
+
+// Has returns true if the endpointSliceTracker has a resource version for the
```
Is this exported to meet interface?

robscott replied:
I think this was originally with the thought that the endpointSliceTracker would move to a util package. I think that's an especially good goal now since it is used by 2 controllers. With that said, that's not going to be part of this PR so it's relatively straightforward to backport. I've changed all the functions here to be unexported.

bowei commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,493 @@
+		servicesByNamespacedName[namespacedName(svc.Namespace, svc.Name)] = svc
+	}
+
+	numEndpointsWithProblems := 0
```
Can we call this `unlabeledCount` to align w/ func name?

robscott replied:
Updated, thanks!

bowei commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,493 @@
+	numEndpointsWithProblems := 0
+	for _, ep := range endpoints {
+		svc, ok := servicesByNamespacedName[namespacedName(ep.Namespace, ep.Name)]
+		if ok && svc.Spec.Selector != nil {
```
Reduce nesting:
`if !ok || svc.Spec.Selector == nil { continue }`

robscott replied:
Updated, thanks!

bowei commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,493 @@
+		if ok && svc.Spec.Selector != nil {
+			labelVal, ok := ep.Labels[discovery.LabelSkipMirror]
+			if !ok || labelVal != "true" {
+				klog.V(4).Infof("Expected %s Endpoints resource to have %s label set to true, got %s", ep.Name, discovery.LabelSkipMirror, labelVal)
```
This is a candidate for V(5) if any

robscott replied:
Moved to v(5). I think this is still valuable as this line can give insight into why the EndpointSliceMirroring controller isn't starting up if that ever happens. Without it, we just have a number of Endpoints, while this gives us actual names of problematic Endpoints. I agree that this could be unnecessarily verbose in most normal startup cases.

robscott replied:
This is now irrelevant (and removed) thanks to Minhan's suggestion to not require the label to be set here.

bowei commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,313 @@
+		pmSlices, pmTotals := r.reconcileByPortMapping(
+			endpoints, existingSlicesByKey[portKey], desiredEndpoints, d.portsByKey[portKey], portKey.addressType())
+
+		totals.added += pmTotals.added
```
```
totals.Add(pmTotals)

func (*totalsByAction) Add(other*totalsByAction)

robscott replied:
Good idea - added.

bowei commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,313 @@
+			Slices:    len(existingSlicesByKey[portKey]) + len(pmSlices.toCreate) - len(pmSlices.toDelete),
+		})
+
+		slices.toCreate = append(slices.toCreate, pmSlices.toCreate...)
```
slices.Append(pmSlices)

robscott replied:
Good idea - added.

bowei commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,313 @@
+			if totalAddresses > int(r.maxEndpointsPerSubset) {
+				break
+			}
+			success := d.addAddress(address, multiKey, true)
```
if ok := d.addAddress(...); !ok { ... }

robscott replied:
Updated, thanks!

bowei commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,313 @@
+			if totalAddresses > int(r.maxEndpointsPerSubset) {
+				break
+			}
+			success := d.addAddress(address, multiKey, true)
```
if ok := d.addAddress(...); !ok

robscott replied:
Updated, thanks!

freehan commented on staging/src/k8s.io/api/discovery/v1beta1/well_known_labels.go:
```diff
@@ -25,4 +25,8 @@ const (
+	// LabelSkipMirror can be set to true on an Endpoints resource to indicate
+	// that the EndpointSliceMirroring controller should not mirror this
+	// resource with EndpointSlices.
+	LabelSkipMirror = "endpointslice.kubernetes.io/skip-mirror"
```
I am wondering if the controller can look up the corresponding service in the cache to see if it has a selector or not, and decide whether it needs to mirror the endpoints. Instead of relying on the label. I guess these are the following scenarios.

1. Service with selector. => no need to mirror
2. Service without selector AND no Endpoints => no need to mirror
3. Service without selector AND has an custom Endpoints object AND no skip mirror label => need to mirror
4. Service without selector AND has an custom Endpoints object AND has skip mirror label => no need to mirror

robscott replied:
This would be a relatively significant deviation from the KEP, but I think it could be worthwhile. This would mean I could remove all changes from the Endpoints controller as part of this PR, and maybe more significantly, would remove the need for the significant startup wait sequence for this controller. If there were any bugs in that startup wait sequence the controller may never actually start up. It looks like @thockin may already [suggested something like this here](https://github.com/kubernetes/enhancements/pull/1713#discussion_r426089134). If there are no objections, I'll start moving towards this approach.

robscott replied:
I've updated this PR to use the suggested approach above, I think it is better than requiring the skip-mirror label to be set everywhere and blocking controller startup for that. One place I have not changed that is the self-referential API Server EndpointSlice. The `kubernetes` EndpointSlice is essentially a mirror of the `kubernetes` Endpoints resource, but it's managed by `pkg/master` and not the EndpointSliceMirroring controller. Changing that to use this controller would be messy at best. It also feels backwards to have API Server be dependent on kube-controller-manager.

aojea replied:
mostly curiosity, is there a case of endpoints without services associated?

robscott replied:
Great question. I can't think of any that are officially supported, though of course it can happen. From what I can tell, kube-proxy will disregard any endpoints that don't have an associated service. (It loops through services and gets any associated endpoints).

aojea replied:
just for clarification, the question was unrelated to this PR :smile:

thockin replied:
Endpoints without a same-name service are automatically removed eventually

freehan commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,497 @@
+	}
+
+	klog.V(2).Infof("Waiting for application of %s label", discovery.LabelSkipMirror)
+	err := c.waitForSkipLabelApplication()
```
This is why I felt like it might be an overkill to apply the label universally. You are already polling the service cache to achieve this.

robscott replied:
I agree with this approach, I've updated the PR to use it.

robscott committed:
Adding endpointslice.kubernetes.io/skip-mirror label

robscott commented:
/retest

aojea commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,306 @@
+			slices.toUpdate = append(slices.toUpdate, slice)
+			slices.toDelete = append(slices.toDelete[:i], slices.toDelete[i+1:]...)
+		} else {
+			i++
```
sorry, dumb question, but I can not understand why `i` it's only incremented here :sweat_smile:

robscott replied:
This is confusing, I'll work on cleaning it up. This loop is modifying the slice it's iterating through inside the loop and only iterates if it doesn't remove something from the slice. Will pull this out so it can be better tested and less confusing.

robscott replied:
Moved this into a separate function and added a bunch of tests - thanks for catching this rather confusing piece of code!

thockin reviewed (commented):
I read it, and I understand it, but I didn't parse it for every possible error.  It seems like you have reviewers doing that.

I'm OK with this after conversation.  Not HAPPY, but OK.

Notes from conversation:

* This is huge.  Did we consider an approach which made consumers deal with this (watch both EP and EPS)?  That's a lot of consumers that have to do the same work - Ingresses, LBs, kube-proxy, kube-proxy replacements, etc.  If we can absorb it, that would be good.

* What happens if I already manually mirrored?  EPS consumers already need to filter for dups, so no big deal.

* Why not filter out Services for which there exists an EPS with the right label?  EPS is better at handling more than less, dups are not a problem, but stale could be.

* This should have had its own KEP.  Yes.  It's unclear from KEP guidelines, but we agree on this one.

* How much of this code is directly copied from EPS controller?  A lot at first, less than half now.

robscott commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,444 @@
+// - the Endpoints resource has a skip-mirror label.
+// - the Endpoints resource has a leader election annotation.
+// - the corresponding Service resource does not exist.
+// - the corresponding Service resource has a nil selector.
```
Fixed, thanks!

robscott commented:
/cc @khenidak

robscott committed:
Adding EndpointSliceMirroring controller

This will mirror custom Endpoints to EndpointSlices to ensure that
applications will not need to maintain both separately.

robscott committed:
Updating apiserver Endpoints management to set skip-mirror label

This will ensure that the self-referential kubernetes Endpoints
resources is not mirrored by the EndpointSliceMirroring controller.

robscott committed:
Adding integration and e2e tests to cover EndpointSliceMirroring

robscott committed:
Enabling the EndpointSliceMirroring controller, adding related config

robscott commented:
/retest

bowei commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,446 @@
+// workqueue guarantees that they will not end up processing the same service
+// at the same time
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
```
// empty loop body

robscott replied:
I think this is just standard for all [Kubernetes controllers](https://github.com/kubernetes/sample-controller/blob/master/controller.go#L181-L182), are you suggesting I add a comment here?

bowei commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,446 @@
+func (c *Controller) syncEndpoints(key string) error {
+	startTime := time.Now()
+	defer func() {
+		syncDuration := float64(time.Since(startTime).Milliseconds()) / float64(1000)
```
do numeric constants need to be typed w/ float?

robscott replied:
Good catch, fixed.

bowei commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,446 @@
+
+	if err != nil {
+		ep := &v1.Endpoints{ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: namespace}}
+		c.eventRecorder.Eventf(ep, "FailedToListEndpointSlices",
```
can we toss the event constants in a file `events.go` (same package)
so these can be documented and are easier to discover?

robscott replied:
Good idea, added that.

bowei commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -0,0 +1,446 @@
+// version in the endpointSliceTracker.
+func (c *Controller) onEndpointSliceDelete(obj interface{}) {
+	endpointSlice := getEndpointSliceFromDeleteAction(obj)
+	if endpointSlice == nil {
```
I thought there is a standard code path where the obj is not of type EndpointSlice, but is some other sentinel value (we've had spurious errors due to this)

We should probably handle that explicitly is that is the case

robscott replied:
That's handled by `getEndpointSliceFromDeleteAction()`. (That function is identical to EndpointSlice controller implementation).

bowei commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,287 @@
+	// Build data structures for existing state.
+	existingSlicesByKey := endpointSlicesByKey(existingSlices)
+
+	// Initialize data structures to track changes.
```
comment is somewhat redundant

robscott replied:
Good call, pulled a comment from below up a bit instead.

bowei commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,287 @@
+	if len(existingSlices) > 0 {
+		newSlice.Name = existingSlices[0].Name
+		slices.toUpdate = []*discovery.EndpointSlice{newSlice}
+	} else {
```
Are we creating the Name somewhere if it is not set here? Is it worth considering setting the Name here to make it easier to understand where the Name value is coming from?

robscott replied:
Reworked this a bit so that name or generateName was always set in `newEndpointSlice()`. Ended up with a few more LOC but I think it might be a bit easier to understand.

bowei commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,287 @@
+
+// finalize creates, updates, and deletes slices as specified
+func (r *reconciler) finalize(endpoints *corev1.Endpoints, slices slicesByAction) error {
+	errs := []error{}
```
var errs []error

robscott replied:
Updated, thanks!

bowei commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,287 @@
+
+	// Don't create more EndpointSlices if corresponding Endpoints resource is
+	// being deleted.
+	if endpoints.DeletionTimestamp == nil {
```
The semantics of this is little weird.

Imagine that the corresponding Endpoint has been marked for deletion and the user scales up their endpoints after that. Those endpoints will randomly be part of the resultant set. We are still updating the endpoint set, but not in its entirety. 

Does it make more sense to not have special behavior for non-Finalized resources? This makes it harder reason about. Are we being too clever here?

robscott replied:
This is mirroring the approach we decided on in https://github.com/kubernetes/kubernetes/pull/91311. The goal here is to not conflict with garbage collection while still supporting existing EndpointSlices in the case deletion takes a long time. In the case of this controller with a 1:1 mapping between EndpointSlice and EndpointSubset, it seems unlikely that an EndpointSlice would be removed for any reason other than garbage collection when there is a non nil deletion timestamp.

bowei commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,287 @@
+	// be deleted.
+	recycleSlices(&slices)
+
+	// Don't create more EndpointSlices if corresponding Endpoints resource is
```
might make below easier to read

`epsClient := r.client.DiscoveryV1beta1().EndpointSlices(endpoints.Namespace)`

robscott replied:
Updated, thanks!

bowei commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,287 @@
+// Endpoints references from the metricsCache.
+func (r *reconciler) deleteEndpoints(namespace, name string, endpointSlices []*discovery.EndpointSlice) error {
+	r.metricsCache.DeleteEndpoints(types.NamespacedName{Namespace: namespace, Name: name})
+	errs := []error{}
```
var errs []error

robscott replied:
Updated, thanks!

bowei commented on pkg/controller/endpointslicemirroring/reconciler_test.go:
```diff
@@ -0,0 +1,914 @@
+	utilpointer "k8s.io/utils/pointer"
+)
+
+var defaultMaxEndpointsPerSubset = int32(1000)
```
const

bowei replied:
Shouldn't we test w/ a tiny size (like 13) to get a better sense of how the slicing code is working?

robscott replied:
Part of the way through this PR I transitioned to an approach that would result in a 1:1 mapping between EndpointSubset and EndpointSlice (potentially 1:2 if there are dual stack IPs). That means there's not a whole lot of actual slicing logic here. With that said, it is a good idea to test with a non default value here so I've added a test covering truncation if the value is set to 2. Also updated the var to a const, thanks!

bowei commented on pkg/controller/endpointslicemirroring/reconciler_test.go:
```diff
@@ -0,0 +1,914 @@
+			continue
+		}
+
+		matchingEndpointsV4 := []discovery.Endpoint{}
```
var matchingEndpointsV4 []discovery.Endpoint

robscott replied:
Updated, thanks!

bowei commented on pkg/controller/endpointslicemirroring/utils.go:
```diff
@@ -0,0 +1,246 @@
+	return epsPorts
+}
+
+// boolPtrChanged returns true if a set of bool pointers have different values.
```
boolPtrEqual would be the more common way to name this

robscott replied:
Updated, thanks!

bowei commented on pkg/controller/endpointslicemirroring/utils.go:
```diff
@@ -0,0 +1,246 @@
+
+// objectRefPtrChanged returns true if a set of object ref pointers have
+// different values.
+func objectRefPtrChanged(ref1, ref2 *corev1.ObjectReference) bool {
```
the "changed" is a bit confusing w/out context -- it's probably better to call this blahEqual

robscott replied:
Updated, thanks!

bowei commented on pkg/controller/endpointslicemirroring/utils.go:
```diff
@@ -0,0 +1,246 @@
+// skipMirror return true if the the LabelSkipMirror label has been set to
+// "true".
+func skipMirror(labels map[string]string) bool {
+	skipMirror, _ := labels[discovery.LabelSkipMirror]
```
you could just do

return skipMirror == "true"

robscott replied:
That is a lot cleaner, thanks!

bowei commented on pkg/controller/endpointslicemirroring/utils.go:
```diff
@@ -0,0 +1,246 @@
+
+// forLeaderElection returns true if the LeaderElectionRecordAnnotationKey is
+// set as an annotation.
+func forLeaderElection(annotations map[string]string) bool {
```
strange name?

hasLeaderElection instead?

robscott replied:
Updated, thanks!

bowei reviewed (commented):
whew, this is long. I think I was able to go through everything.

robscott commented:
/retest

robscott commented:
/retest

thockin commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,291 @@
+// desired state
+type reconciler struct {
+	client                clientset.Interface
+	maxEndpointsPerSubset int32
```
Please - some comments on things like this would help :)

robscott replied:
Good call, I had some comments about this elsewhere but completely missed this place. I've added better comments for the surrounding reconciler fields as well.

thockin commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -0,0 +1,291 @@
+
+		for _, address := range subset.Addresses {
+			totalAddresses++
+			if totalAddresses > int(r.maxEndpointsPerSubset) {
```
I'd love to see some comments explaining this truncation as intentional and why

thockin replied:
Should we have a metric for how many endpoints we DIDN'T mirror?

And log lines?

robscott replied:
Good call, thanks! I've added comments, logging, an event, and metrics around this now.

thockin reviewed (commented):
I'm leaning a lot on the other reviewers here.  My main concern is that there are some things that are not obvious (potentially surprising to users) and not well explained.  I spent a bunch of time understanding a "bug" that turns out to have been intentional.

Im going to go ahead an /approve it, but I'd love to see some of the more subtle parts explained/commented more.  Maybe as one last commit.

/approve

robscott committed:
Adding new logging, event, and metric to better capture when mirroring addresses is skipped

robscott commented:
/retest

robscott commented:
/retest

robscott commented:
/retest

freehan reviewed (approved):
/lgtm
/approve

aojea commented:
/retest

robscott commented:
/retest

robscott commented:
/retest

bowei commented:
/lgtm

robscott commented:
/retest

robscott commented:
/retest

robscott commented:
/retest

robscott commented:
/retest

Antiarchitect commented:
Will this be backported to 1.18? Currently facing this issue with Cilium 1.8.2 and K8s 1.18.8 with services without selector.

robscott commented:
@Antiarchitect Is that with the EndpointSliceProxying feature gate enabled on kube-proxy? Although I'd love to cherry pick this I think it's likely too big to pull back to a previous Kubernetes version.

Antiarchitect commented:
@robscott There is no kube-proxy with Cilium - it provides a full replacement for it.

robscott commented:
@Antiarchitect woops, I missed that detail. The assumption in 1.18 and earlier was that anything publishing custom Endpoints would starting publishing custom EndpointSlices. That ended up being naive so we added the EndpointSliceMirroring controller in 1.19. Unfortunately any use of EndpointSlices in 1.18 by a proxy is going to require any custom Endpoints to be replicated to EndpointSlices through some other means.

Antiarchitect commented:
Actually there is a way in Cilium to cut the consequences: https://github.com/cilium/cilium/issues/12513#issuecomment-658260758 I'll stick with that until I upgrade to K8s 1.19.
